### PR TITLE
implement filter for the reload layer/subscriber

### PR DESCRIPTION
**Pull request by guswynn**
## Motivation
The `reload` layer doesn't (and can't) implement downcasting correctly, which breaks certain subscribers like the opentelemetry one.
## Solution

. Most usages of this type (including mine) are just to change the filter, so this pr implements `Filter` on the reload type to allow users to not need to wrap the whole layer trait. Another advantage of this is that the common-case critical sections are shortened

### CODE REVIEW

File: tracing-subscriber/src/reload.rs
Comment (hawkw) [inline tracing-subscriber/src/reload.rs:185]: nit: why `O`? does that stand for something?

**Review by hawkw — APPROVED** [review]
looks good to me overall, thanks for adding a test! i had some minor suggestions.

**Comment by guswynn** [comment]
@hawkw thanks for the review! This is RFAL!

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:220]:
```suggestion
    /// **Warning:** [`Handle::reload`] cannot be used with the [`Filtered`](crate::filter::Filtered)
    /// subscriber; use [`Handle::modify`] instead (this [this issue] for additional details).
```

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:223]:
```suggestion
    /// However, if the _only_ the [`Filter`] needs to be modified, use `reload::Subscriber` to
    /// to wrap the `Filter` directly.
    /// 
    /// [this issue]: https://github.com/tokio-rs/tracing/issues/1629
```

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:224]:
```suggestion
    pub fn reload(&self, new_value: impl Into<T>) -> Result<(), Error> {
```

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:268]:
```suggestion
    pub fn with_current<T2>(&self, f: impl FnOnce(&T) -> T2) -> Result<T2, Error> {
```

(we make use of `T` and `T2` in other parts of `tracing-subscriber`.)

**Review by davidbarsky — CHANGES_REQUESTED** [review]
made a few requested changes as part of our adherence to our style guide. broadly speaking, we don't use the second-person tense in reference documentation, but do use it in longer-term guides (e.g., https://tokio.rs/tokio/tutorial). I apologize for being a stickler for this stuff and I wish we can automate this better :)

**Comment by davidbarsky** [comment]
> ...we should probably actually write down a style guide for docs, huh? maybe in `CONTRIBUTING.md`?

Probably. An example of some documentation on a method, a struct, and some do's and do not's would probably go a long way.

File: tracing-subscriber/src/reload.rs
Comment (guswynn) [reply tracing-subscriber/src/reload.rs:220]: I feel like I should still mention wrapping the filter?

**Comment by guswynn** [comment]
RFAL! I feel like I should re-export `reload::Subscriber` as `reload::Filter` as well...

**Comment by hawkw** [comment]
> RFAL! I feel like I should re-export `reload::Subscriber` as `reload::Filter` as well, or name it something else and re-export as `Subscriber`

instead of re-exporting it under two different names, i would probably just rename it to something more generic (like `reload::Reload`) and deprecate the existing name. Let's do that in a follow-up, though, because we can just do a hard rename on v0.2.x, and we'll need to do a deprecation on v0.1.x.

**Comment by guswynn** [comment]
@hawkw okay! after this merges Ill do that rename!

**Comment by hawkw** [comment]
@davidbarsky do you want to review the docs again, or shall i merge this?

**Comment by davidbarsky** [comment]
> @davidbarsky do you want to review the docs again, or shall i merge this?

I realized I missed a bit on the module documentation; lemme fix that. I think it should be good to go after, though!

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:19]: /// [`Handle::reload`] cannot be used with the [`Filtered`](crate::filter::Filtered)
    /// subscriber; use [`Handle::modify`] instead (see [this issue] for additional details).
    ///
    /// However, if the _only_ the [`Filter`](crate::subscribe::Filter) needs to be modified,
    /// use `reload::Subscriber` to wrap the `Filter` directly.
    ///
    /// [this issue]: https://github.com/tokio-rs/tracing/issues/1629

File: tracing-subscriber/src/reload.rs
Comment (davidbarsky) [inline tracing-subscriber/src/reload.rs:217]:
```suggestion
    /// **Warning:** [`Handle::reload`] cannot be used with the [`Filtered`](crate::filter::Filtered)
    /// subscriber; use [`Handle::modify`] instead (see [this issue] for additional details).
```

**Review by davidbarsky — COMMENTED** [review]
missed a few things, sorry!

File: tracing-subscriber/src/reload.rs
Comment (guswynn) [reply tracing-subscriber/src/reload.rs:19]: removing the `Warning`?

**Comment by guswynn** [comment]
Is this merge-able?

**Comment by hawkw** [comment]
> Is this merge-able?

Yup, I'll get it merged and then backport to v0.1.x!

**Pull request merged**
